diff -Naur gtk+-2.12.2/README.xynth gtk+-2.12.2-xynth/README.xynth
--- gtk+-2.12.2/README.xynth	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.12.2-xynth/README.xynth	2007-11-30 15:19:42.000000000 +0200
@@ -0,0 +1,7 @@
+export PKG_CONFIG_PATH=/opt/gtk/lib/pkgconfig
+export LD_LIBRARY_PATH=/opt/gtk/lib
+aclocal && automake && autoconf
+mkdir build-xynth
+cd build-xynth
+../configure --with-gdktarget=xynth --prefix=/opt/gtk --disable-visibility
+make
diff -Naur gtk+-2.12.2/configure.in gtk+-2.12.2-xynth/configure.in
--- gtk+-2.12.2/configure.in	2007-11-26 21:40:52.000000000 +0200
+++ gtk+-2.12.2-xynth/configure.in	2007-11-30 15:22:10.000000000 +0200
@@ -239,13 +239,13 @@
   gdktarget=x11
 fi
 
-AC_ARG_WITH(gdktarget, [  --with-gdktarget=[[x11/win32/quartz/directfb]] select non-default GDK target],
+AC_ARG_WITH(gdktarget, [  --with-gdktarget=[[x11/win32/quartz/directfb/xynth]] select non-default GDK target],
 	gdktarget=$with_gdktarget)
 
 AC_SUBST(gdktarget)
 case $gdktarget in
-  x11|win32|quartz|directfb) ;;
-  *) AC_MSG_ERROR([Invalid target for GDK: use x11, quartz, directfb or win32.]);;
+  x11|win32|quartz|directfb|xynth) ;;
+  *) AC_MSG_ERROR([Invalid target for GDK: use x11, quartz, directfb, xynth or win32.]);;
 esac
 
 gdktargetlib=libgdk-$gdktarget-$GTK_API_VERSION.la
@@ -1418,6 +1418,26 @@
   AM_CONDITIONAL(USE_DIRECTFB, false)
 fi
 
+if test "x$gdktarget" = "xxynth"; then
+  XYNTH_REQUIRED_VERSION=0.8.40
+  AC_MSG_CHECKING(for Xynth)
+
+  if $PKG_CONFIG --atleast-version $XYNTH_REQUIRED_VERSION xynth ; then
+      AC_MSG_RESULT(found)
+      GDK_EXTRA_CFLAGS="`$PKG_CONFIG --cflags xynth` $GDK_EXTRA_CFLAGS"
+      GDK_EXTRA_LIBS="`$PKG_CONFIG --libs xynth` $GDK_EXTRA_LIBS"
+  else
+      AC_MSG_ERROR([
+*** Xynth $XYNTH_REQUIRED_VERSION or newer and the cairo backend 
+*** are required. The latest version of Xynth is always available 
+*** from http://www.xynth.org/.
+  ])
+  fi
+
+  AM_CONDITIONAL(USE_XYNTH, true)
+else
+  AM_CONDITIONAL(USE_XYNTH, false)
+fi
 
 # Check for Pango flags
 
@@ -1722,6 +1742,9 @@
 elif test "x$gdktarget" = "xdirectfb" ; then
   gdk_windowing='
 #define GDK_WINDOWING_DIRECTFB'
+elif test "x$gdktarget" = "xxynth" ; then
+  gdk_windowing='
+#define GDK_WINDOWING_XYNTH'
 fi
 
 if test x$gdk_wchar_h = xyes; then
@@ -1780,6 +1803,7 @@
 gdk/win32/rc/gdk.rc
 gdk/quartz/Makefile
 gdk/directfb/Makefile
+gdk/xynth/Makefile
 gtk/Makefile
 gtk/makefile.msc
 gtk/gtkversion.h
diff -Naur gtk+-2.12.2/gdk/Makefile.am gtk+-2.12.2-xynth/gdk/Makefile.am
--- gtk+-2.12.2/gdk/Makefile.am	2007-11-26 21:40:52.000000000 +0200
+++ gtk+-2.12.2-xynth/gdk/Makefile.am	2007-11-30 15:23:00.000000000 +0200
@@ -1,7 +1,7 @@
 ## Makefile.am for gtk+/gdk
 
 SUBDIRS = $(gdktarget)
-DIST_SUBDIRS = win32 x11 quartz directfb
+DIST_SUBDIRS = win32 x11 quartz directfb xynth
 
 EXTRA_DIST =			\
 	keynames.txt		\
@@ -142,6 +142,11 @@
 	gdkenumtypes.c		\
 	gdkmarshalers.h
 
+libgdk_xynth_2_0_la_SOURCES = $(common_sources) 
+libgdk_xynth_2_0_la_LIBADD = xynth/libgdk-xynth.la $(GDK_DEP_LIBS) \
+  $(top_builddir)/gdk-pixbuf/libgdk_pixbuf-$(GTK_API_VERSION).la
+libgdk_xynth_2_0_la_LDFLAGS = $(LDADD)
+
 libgdk_directfb_2_0_la_SOURCES = $(common_sources) 
 libgdk_directfb_2_0_la_LIBADD = directfb/libgdk-directfb.la $(GDK_DEP_LIBS) \
   $(top_builddir)/gdk-pixbuf/libgdk_pixbuf-$(GTK_API_VERSION).la
@@ -205,7 +210,7 @@
 
 lib_LTLIBRARIES = $(gdktargetlib)
 
-EXTRA_LTLIBRARIES = libgdk-x11-2.0.la libgdk-win32-2.0.la libgdk-quartz-2.0.la libgdk-directfb-2.0.la
+EXTRA_LTLIBRARIES = libgdk-x11-2.0.la libgdk-win32-2.0.la libgdk-quartz-2.0.la libgdk-directfb-2.0.la libgdk-xynth-2.0.la
 
 MAINTAINERCLEANFILES = gdkenumtypes.h stamp-gdkenumtypes.h gdkenumtypes.c \
   gdkmarshalers.h gdkmarshalers.c
diff -Naur gtk+-2.12.2/gdk/xynth/Makefile.am gtk+-2.12.2-xynth/gdk/xynth/Makefile.am
--- gtk+-2.12.2/gdk/xynth/Makefile.am	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.12.2-xynth/gdk/xynth/Makefile.am	2007-11-30 15:24:13.000000000 +0200
@@ -0,0 +1,24 @@
+## Makefile.am for gtk+/gdk/gdk-xynth
+
+libgdkincludedir = $(includedir)/gtk-2.0/gdk
+
+INCLUDES = 				\
+	-DG_LOG_DOMAIN=\"Gdk-Xynth\"	\
+	-DGDK_COMPILATION		\
+	-I$(top_srcdir)			\
+	-I$(top_srcdir)/gdk		\
+	-I$(top_builddir)/gdk		\
+	-DG_DISABLE_DEPRECATED		\
+	@GTK_DEBUG_FLAGS@		\
+	@GDK_DEP_CFLAGS@		
+
+LDADDS = 		\
+	@GDK_DEP_LIBS@		
+
+noinst_LTLIBRARIES = libgdk-xynth.la
+
+libgdk_xynth_la_SOURCES =	\
+	gdkxynth.c
+
+libgdkinclude_HEADERS =		\
+	gdkxynth.h
diff -Naur gtk+-2.12.2/gdk/xynth/gdkxynth.c gtk+-2.12.2-xynth/gdk/xynth/gdkxynth.c
--- gtk+-2.12.2/gdk/xynth/gdkxynth.c	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.12.2-xynth/gdk/xynth/gdkxynth.c	2007-11-30 15:26:59.000000000 +0200
@@ -0,0 +1,2 @@
+
+#incude "gdkxynth.h"
diff -Naur gtk+-2.12.2/gdk/xynth/gdkxynth.h gtk+-2.12.2-xynth/gdk/xynth/gdkxynth.h
--- gtk+-2.12.2/gdk/xynth/gdkxynth.h	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.12.2-xynth/gdk/xynth/gdkxynth.h	2007-11-30 15:27:03.000000000 +0200
@@ -0,0 +1 @@
+
diff -Naur gtk+-2.12.2/gtk/Makefile.am gtk+-2.12.2-xynth/gtk/Makefile.am
--- gtk+-2.12.2/gtk/Makefile.am	2007-11-26 21:40:15.000000000 +0200
+++ gtk+-2.12.2-xynth/gtk/Makefile.am	2007-11-30 15:20:50.000000000 +0200
@@ -831,12 +831,14 @@
 libgtk_win32_2_0_la_SOURCES = $(gtk_c_sources)
 libgtk_quartz_2_0_la_SOURCES = $(gtk_c_sources)
 libgtk_directfb_2_0_la_SOURCES = $(gtk_c_sources)
+libgtk_xynth_2_0_la_SOURCES = $(gtk_c_sources)
 
 libgtk_x11_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_linux_fb_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_win32_2_0_la_LDFLAGS = $(libtool_opts) -Wl,-luuid
 libgtk_quartz_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_directfb_2_0_la_LDFLAGS = $(libtool_opts)
+libgtk_xynth_2_0_la_LDFLAGS = $(libtool_opts)
 
 libgtk_x11_2_0_la_LIBADD = $(libadd)
 libgtk_linux_fb_2_0_la_LIBADD = $(libadd)
@@ -844,11 +846,12 @@
 libgtk_win32_2_0_la_DEPENDENCIES = $(gtk_def) $(gtk_win32_res)
 libgtk_quartz_2_0_la_LIBADD = $(libadd)
 libgtk_directfb_2_0_la_LIBADD = $(libadd)
+libgtk_xynth_2_0_la_LIBADD = $(libadd)
 
 if USE_WIN32
 libgtk_target_ldflags = $(gtk_win32_res_ldflag) $(gtk_win32_symbols)
 endif
-EXTRA_LTLIBRARIES = libgtk-x11-2.0.la libgtk-linux-fb-2.0.la libgtk-win32-2.0.la libgtk-quartz-2.0.la libgtk-directfb-2.0.la
+EXTRA_LTLIBRARIES = libgtk-x11-2.0.la libgtk-linux-fb-2.0.la libgtk-win32-2.0.la libgtk-quartz-2.0.la libgtk-directfb-2.0.la libgtk-xynth-2.0.la
 
 install-exec-hook: 
 if DISABLE_EXPLICIT_DEPS
