diff -Naur gtk+-2.10.14/README.xynth gtk+-2.10.14-xynth/README.xynth
--- gtk+-2.10.14/README.xynth	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.10.14-xynth/README.xynth	2007-12-03 17:18:32.000000000 +0200
@@ -0,0 +1,7 @@
+export PKG_CONFIG_PATH=/opt/gtk/lib/pkgconfig
+export LD_LIBRARY_PATH=/opt/gtk/lib
+aclocal && automake && autoconf
+mkdir build-xynth
+cd build-xynth
+../configure --with-gdktarget=xynth --prefix=/opt/gtk --disable-visibility
+make
diff -Naur gtk+-2.10.14/configure.in gtk+-2.10.14-xynth/configure.in
--- gtk+-2.10.14/configure.in	2007-07-16 22:46:25.000000000 +0300
+++ gtk+-2.10.14-xynth/configure.in	2007-12-03 17:18:32.000000000 +0200
@@ -224,13 +224,13 @@
   gdktarget=x11
 fi
 
-AC_ARG_WITH(gdktarget, [  --with-gdktarget=[[x11/linux-fb/win32/quartz/directfb]] select non-default GDK target],
+AC_ARG_WITH(gdktarget, [  --with-gdktarget=[[x11/linux-fb/win32/quartz/directfb/xynth]] select non-default GDK target],
 	gdktarget=$with_gdktarget)
 
 AC_SUBST(gdktarget)
 case $gdktarget in
-  x11|linux-fb|win32|quartz|directfb) ;;
-  *) AC_MSG_ERROR([Invalid target for GDK: use x11, linux-fb, quartz, directfb or win32.]);;
+  x11|linux-fb|win32|quartz|directfb|xynth) ;;
+  *) AC_MSG_ERROR([Invalid target for GDK: use x11, linux-fb, quartz, directfb, win32 or xynth.]);;
 esac
 
 gdktargetlib=libgdk-$gdktarget-$GTK_API_VERSION.la
@@ -1437,6 +1437,26 @@
   AM_CONDITIONAL(USE_DIRECTFB, false)
 fi
 
+if test "x$gdktarget" = "xxynth"; then
+  XYNTH_REQUIRED_VERSION=0.8.40
+  AC_MSG_CHECKING(for Xynth)
+
+  if $PKG_CONFIG --atleast-version $XYNTH_REQUIRED_VERSION xynth ; then
+      AC_MSG_RESULT(found)
+      GDK_EXTRA_CFLAGS="`$PKG_CONFIG --cflags xynth` $GDK_EXTRA_CFLAGS"
+      GDK_EXTRA_LIBS="`$PKG_CONFIG --libs xynth` $GDK_EXTRA_LIBS"
+  else
+      AC_MSG_ERROR([
+*** Xynth $XYNTH_REQUIRED_VERSION or newer and the cairo backend 
+*** are required. The latest version of Xynth is always available 
+*** from http://www.xynth.org/.
+  ])
+  fi
+
+  AM_CONDITIONAL(USE_XYNTH, true)
+else
+  AM_CONDITIONAL(USE_XYNTH, false)
+fi
 
 # Check for Pango flags
 
@@ -1703,6 +1723,9 @@
 elif test "x$gdktarget" = "xdirectfb" ; then
   gdk_windowing='
 #define GDK_WINDOWING_DIRECTFB'
+elif test "x$gdktarget" = "xxynth" ; then
+  gdk_windowing='
+#define GDK_WINDOWING_XYNTH'
 fi
 
 if test x$gdk_wchar_h = xyes; then
@@ -1762,6 +1785,7 @@
 gdk/linux-fb/Makefile
 gdk/quartz/Makefile
 gdk/directfb/Makefile
+gdk/xynth/Makefile
 gtk/Makefile
 gtk/makefile.msc
 gtk/gtkversion.h
diff -Naur gtk+-2.10.14/gdk/Makefile.am gtk+-2.10.14-xynth/gdk/Makefile.am
--- gtk+-2.10.14/gdk/Makefile.am	2007-07-16 22:46:24.000000000 +0300
+++ gtk+-2.10.14-xynth/gdk/Makefile.am	2007-12-03 17:18:32.000000000 +0200
@@ -1,7 +1,7 @@
 ## Makefile.am for gtk+/gdk
 
 SUBDIRS = $(gdktarget)
-DIST_SUBDIRS = linux-fb win32 x11 quartz directfb
+DIST_SUBDIRS = linux-fb win32 x11 quartz directfb xynth
 
 EXTRA_DIST =			\
 	keynames.txt		\
@@ -134,6 +134,11 @@
 	gdkenumtypes.c		\
 	gdkmarshalers.h
 
+libgdk_xynth_2_0_la_SOURCES = $(common_sources) 
+libgdk_xynth_2_0_la_LIBADD = xynth/libgdk-xynth.la $(GDK_DEP_LIBS) \
+  $(top_builddir)/gdk-pixbuf/libgdk_pixbuf-$(GTK_API_VERSION).la
+libgdk_xynth_2_0_la_LDFLAGS = $(LDADD)
+
 libgdk_directfb_2_0_la_SOURCES = $(common_sources) 
 libgdk_directfb_2_0_la_LIBADD = directfb/libgdk-directfb.la $(GDK_DEP_LIBS) \
   $(top_builddir)/gdk-pixbuf/libgdk_pixbuf-$(GTK_API_VERSION).la
@@ -208,7 +213,7 @@
 
 lib_LTLIBRARIES = $(gdktargetlib)
 
-EXTRA_LTLIBRARIES = libgdk-x11-2.0.la libgdk-linux-fb-2.0.la libgdk-win32-2.0.la libgdk-quartz-2.0.la libgdk-directfb-2.0.la
+EXTRA_LTLIBRARIES = libgdk-x11-2.0.la libgdk-linux-fb-2.0.la libgdk-win32-2.0.la libgdk-quartz-2.0.la libgdk-directfb-2.0.la libgdk-xynth-2.0.la
 
 MAINTAINERCLEANFILES = gdkenumtypes.h stamp-gdkenumtypes.h gdkenumtypes.c \
   gdkmarshalers.h gdkmarshalers.c
diff -Naur gtk+-2.10.14/gdk/xynth/Makefile.am gtk+-2.10.14-xynth/gdk/xynth/Makefile.am
--- gtk+-2.10.14/gdk/xynth/Makefile.am	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.10.14-xynth/gdk/xynth/Makefile.am	2007-12-03 13:11:26.000000000 +0200
@@ -0,0 +1,24 @@
+## Makefile.am for gtk+/gdk/gdk-xynth
+
+libgdkincludedir = $(includedir)/gtk-2.0/gdk
+
+INCLUDES = 				\
+	-DG_LOG_DOMAIN=\"Gdk-Xynth\"	\
+	-DGDK_COMPILATION		\
+	-I$(top_srcdir)			\
+	-I$(top_srcdir)/gdk		\
+	-I$(top_builddir)/gdk		\
+	-DG_DISABLE_DEPRECATED		\
+	@GTK_DEBUG_FLAGS@		\
+	@GDK_DEP_CFLAGS@		
+
+LDADDS = 		\
+	@GDK_DEP_LIBS@		
+
+noinst_LTLIBRARIES = libgdk-xynth.la
+
+libgdk_xynth_la_SOURCES =	\
+	gdkxynth.c
+
+libgdkinclude_HEADERS =		\
+	gdkxynth.h
diff -Naur gtk+-2.10.14/gdk/xynth/gdkxynth.c gtk+-2.10.14-xynth/gdk/xynth/gdkxynth.c
--- gtk+-2.10.14/gdk/xynth/gdkxynth.c	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.10.14-xynth/gdk/xynth/gdkxynth.c	2007-12-03 17:14:57.000000000 +0200
@@ -0,0 +1,1529 @@
+
+#include "gdkxynth.h"
+
+const GOptionEntry _gdk_windowing_args[] = {
+	{ NULL }
+};
+
+void gdk_window_show (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_text_property_to_utf8_list_for_display (GdkDisplay *display, GdkAtom encoding, gint format, const guchar *text, gint length, gchar ***list)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_event_send_client_message_for_display (GdkDisplay *display, GdkEvent *event, guint32 xid)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_display_supports_cursor_alpha (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkWindow * gdk_selection_owner_get_for_display (GdkDisplay *display, GdkAtom selection)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkAtom gdk_atom_intern (const gchar *atom_name, gboolean only_if_exists)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_drop_finish (GdkDragContext *context, gboolean success, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_screen_broadcast_client_message (GdkScreen *screen, GdkEvent *sev)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_screen_get_height (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_device_set_key (GdkDevice *device, guint index, guint keyval, GdkModifierType modifiers)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_screen_get_setting (GdkScreen *screen, const gchar *name, GValue *value)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_keymap_translate_keyboard_state (GdkKeymap *keymap, guint keycode, GdkModifierType state, gint group, guint *keyval, gint *effective_group, gint *level, GdkModifierType *consumed_modifiers)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_display_pointer_is_grabbed (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkEvent * gdk_event_get_graphics_expose (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_focus (GdkWindow *window, guint32 timestamp)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GType gdk_device_get_type (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_property_get (GdkWindow *window, GdkAtom property, GdkAtom type, gulong offset, gulong length, gint pdelete, GdkAtom *actual_property_type, gint *actual_format_type, gint *actual_length, guchar **data)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GList * gdk_display_list_devices (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkGrabStatus gdk_keyboard_grab (GdkWindow *window, gboolean owner_events, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gchar * gdk_atom_name (GdkAtom atom)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_configure_finished (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_scroll (GdkWindow *window, gint dx, gint dy)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_window_get_origin (GdkWindow *window, gint *x, gint *y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_skip_pager_hint (GdkWindow *window, gboolean skips_pager)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_device_set_axis_use (GdkDevice *device, guint index, GdkAxisUse use)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkWindow * gdk_window_new (GdkWindow *parent, GdkWindowAttr *attributes, gint attributes_mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gchar * gdk_set_locale (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkGrabStatus gdk_pointer_grab (GdkWindow *window, gboolean owner_events, GdkEventMask event_mask, GdkWindow *confine_to, GdkCursor *cursor, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkCursor * gdk_cursor_new_from_name (GdkDisplay  *display, const gchar *name)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_unmaximize (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_drag_abort (GdkDragContext *context, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_colormap_query_color (GdkColormap *colormap, gulong pixel, GdkColor *result)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_stick (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_selection_owner_set_for_display (GdkDisplay *display, GdkWindow *owner, GdkAtom selection, guint32 time, gboolean send_event)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_display_supports_clipboard_persistence (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gchar * gdk_utf8_to_string_target (const gchar *str)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_device_get_axis (GdkDevice *device, gdouble *axes, GdkAxisUse use, gdouble *value)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_screen_get_number (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GType gdk_image_get_type (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_cursor (GdkWindow *window, GdkCursor *cursor)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkCursor * gdk_cursor_new_from_pixbuf (GdkDisplay *display, GdkPixbuf *pixbuf, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_move (GdkWindow *window, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkPixmap * gdk_pixmap_new (GdkDrawable *drawable, gint width, gint height, gint depth)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_drag_drop (GdkDragContext *context, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_functions (GdkWindow *window, GdkWMFunction functions)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkAtom gdk_drag_get_selection (GdkDragContext *context)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_screen_get_width (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GType gdk_drag_context_get_type (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkDragContext * gdk_drag_begin (GdkWindow *window, GList *targets)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_maximize (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_free_compound_text (guchar *ctext)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_display_set_sm_client_id (GdkDisplay *display, const gchar *sm_client_id)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint _gdk_windowing_get_bits_for_depth (GdkDisplay *display, gint depth)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void _gdk_windowing_get_pointer (GdkDisplay *display, GdkScreen **screen, gint *x, gint *y, GdkModifierType  *mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_window_clear_area (GdkWindow *window, gint x, gint y, gint width, gint height)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GType _gdk_window_impl_get_type (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void _gdk_windowing_set_default_display (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_window_destroy (GdkWindow *window, gboolean recursing, gboolean foreign_destroy)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_withdraw (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_urgency_hint (GdkWindow *window, gboolean urgent)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GList * gdk_screen_list_visuals (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_display_flush (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_accept_focus (GdkWindow *window, gboolean accept_focus)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_screen_get_monitor_geometry (GdkScreen *screen, gint monitor_num, GdkRectangle *dest)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkColormap * gdk_colormap_new (GdkVisual *visual, gboolean allocate)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void _gdk_cursor_destroy (GdkCursor *cursor)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_property_delete (GdkWindow *window, GdkAtom property)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_deiconify (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_font_equal (const GdkFont *fonta, const GdkFont *fontb)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_register_dnd (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_keep_below (GdkWindow *window, gboolean setting)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_selection_convert (GdkWindow *requestor, GdkAtom selection, GdkAtom target, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_device_set_mode (GdkDevice *device, GdkInputMode mode)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_fullscreen (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_display_sync (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_text_width (GdkFont *font, const gchar *text, gint text_length)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkWindow * gdk_screen_get_root_window (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_mbstowcs (GdkWChar *dest, const gchar *src, gint dest_max)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_unstick (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_window_get_offsets (GdkWindow *window, gint *x_offset, gint *y_offset)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_display_supports_cursor_color (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_screen_is_composited (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_iconify (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_selection_send_notify_for_display (GdkDisplay *display, guint32 requestor, GdkAtom selection, GdkAtom target, GdkAtom property, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_input_set_extension_events (GdkWindow *window, gint mask, GdkExtensionMode mode)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+guint32 gdk_drag_get_protocol_for_display (GdkDisplay *display, guint32 xid, GdkDragProtocol *protocol)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_error_trap_push (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_get_geometry (GdkWindow *window, gint *x, gint *y, gint *width, gint *height, gint *depth)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkAtom gdk_atom_intern_static_string (const gchar *atom_name)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void _gdk_windowing_init (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void  gdk_window_set_geometry_hints (GdkWindow *window, GdkGeometry *geometry, GdkWindowHints geom_mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_display_beep (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkColormap * gdk_screen_get_system_colormap (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void  gdk_device_get_state (GdkDevice *device, GdkWindow *window, gdouble *axes, GdkModifierType *mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_utf8_to_compound_text_for_display (GdkDisplay *display, const gchar *str, GdkAtom *encoding, gint *format, guchar **ctext, gint *length)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkWindow * _gdk_windowing_window_get_pointer (GdkDisplay *display, GdkWindow *window, gint *x, gint *y, GdkModifierType *mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_shape_combine_mask (GdkWindow *window, GdkBitmap *mask, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_back_pixmap (GdkWindow *window, GdkPixmap *pixmap, gboolean parent_relative)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_events_queue (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_window_clear_area_e (GdkWindow *window, gint x, gint y, gint width, gint height)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_window_destroy_foreign (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkFont * gdk_font_load_for_display (GdkDisplay *display, const gchar *font_name)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_screen_set_default_colormap (GdkScreen *screen, GdkColormap *colormap)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_drag_motion (GdkDragContext *context, GdkWindow *dest_window, GdkDragProtocol protocol, gint x_root, gint y_root, GdkDragAction suggested_action, GdkDragAction possible_actions, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkPixbuf * gdk_cursor_get_image (GdkCursor *cursor)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_drag_find_window_for_screen (GdkDragContext *context, GdkWindow *drag_window, GdkScreen *screen, gint x_root, gint y_root, GdkWindow **dest_window, GdkDragProtocol *protocol)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+guint32 gdk_image_get_pixel (GdkImage *image, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_display_store_clipboard (GdkDisplay *display, GdkWindow *clipboard_window, guint32 time_, GdkAtom *targets, gint n_targets)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_pointer_grab_info_libgtk_only (GdkDisplay *display, GdkWindow **grab_window, gboolean *owner_events)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkScreen * gdk_visual_get_screen (GdkVisual *visual)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkScreen * gdk_display_get_screen (GdkDisplay *display, gint screen_num)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_text_property_to_text_list_for_display (GdkDisplay *display, GdkAtom encoding, gint format, const guchar *text, gint length, gchar ***list)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_string_to_compound_text_for_display (GdkDisplay *display, const gchar *str, GdkAtom *encoding, gint *format, guchar **ctext, gint *length)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_error_trap_pop (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkWindow * _gdk_windowing_window_at_pointer (GdkDisplay *display, gint *win_x, gint *win_y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void _gdk_font_destroy (GdkFont *font)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void _gdk_windowing_gc_copy (GdkGC *dst_gc, GdkGC *src_gc)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_keyboard_grab_info_libgtk_only (GdkDisplay *display, GdkWindow **grab_window, gboolean *owner_events)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GType gdk_colormap_get_type (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_display_warp_pointer (GdkDisplay *display, GdkScreen *screen, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_modal_hint (GdkWindow *window, gboolean modal)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_font_id (const GdkFont *font)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_raise (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+guint gdk_keyval_from_name (const gchar *keyval_name)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkDisplay * gdk_screen_get_display (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_reparent (GdkWindow *window, GdkWindow *new_parent, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkPixmap * gdk_bitmap_create_from_data (GdkDrawable *drawable, const gchar *data, gint width, gint height)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkCursor * gdk_cursor_new_for_display (GdkDisplay *display, GdkCursorType cursor_type)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkVisual * gdk_screen_get_system_visual (GdkScreen * screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_screen_get_width_mm (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_text_extents (GdkFont *font, const gchar *text, gint text_length, gint *lbearing, gint *rbearing, gint *width, gint *ascent, gint *descent)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_get_frame_extents (GdkWindow *window, GdkRectangle *rect)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_window_get_deskrelative_origin (GdkWindow *window, gint *x, gint *y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gchar * gdk_wcstombs (const GdkWChar *src)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_screen_get_n_monitors (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkScreen * gdk_display_get_default_screen (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_transient_for (GdkWindow *window, GdkWindow *parent)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_text_width_wc (GdkFont *font, const GdkWChar *text, gint text_length)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkDisplay * gdk_cursor_get_display (GdkCursor *cursor)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void _gdk_windowing_gc_set_clip_region (GdkGC *gc, GdkRegion *region)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_events (GdkWindow *window, GdkEventMask event_mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_role (GdkWindow *window, const gchar *role)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_decorations (GdkWindow *window, GdkWMDecoration decorations)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_show_unraised (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_colormap_change (GdkColormap *colormap, gint ncolors)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_property_change (GdkWindow *window, GdkAtom property, GdkAtom type, gint format, GdkPropMode mode, const guchar *data, gint nelements)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_begin_resize_drag (GdkWindow *window, GdkWindowEdge edge, gint button, gint root_x, gint root_y, guint32 timestamp)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_colormap_free_colors (GdkColormap *colormap, GdkColor *colors, gint ncolors)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_enable_synchronized_configure (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_focus_on_map (GdkWindow *window, gboolean focus_on_map)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_unfullscreen (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_skip_taskbar_hint (GdkWindow *window, gboolean skips_taskbar)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkImage * _gdk_image_new_for_depth (GdkScreen *screen, GdkImageType type, GdkVisual *visual, gint width, gint height, gint depth)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint _gdk_font_strlen (GdkFont *font, const gchar *str)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_selection_property_get (GdkWindow *requestor, guchar **data, GdkAtom *ret_type, gint *ret_format)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gchar * _gdk_windowing_substitute_screen_number (const gchar *display_name, gint screen_number)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkCursor * gdk_cursor_new_from_pixmap (GdkPixmap *source, GdkPixmap *mask, const GdkColor *fg, const GdkColor *bg, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_flush (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+PangoDirection gdk_keymap_get_direction (GdkKeymap *keymap)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_title (GdkWindow *window, const gchar *title)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkScreen * gdk_colormap_get_screen (GdkColormap *cmap)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_hide (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_notify_startup_complete (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_display_get_maximal_cursor_size (GdkDisplay *display, guint *width, guint *height)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkWindow * gdk_window_foreign_new_for_display (GdkDisplay *display, GdkNativeWindow anid)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_type_hint (GdkWindow *window, GdkWindowTypeHint hint)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkEventMask gdk_window_get_events (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_keep_above (GdkWindow *window, gboolean setting)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void  gdk_window_input_shape_combine_mask (GdkWindow *window, GdkBitmap *mask, gint x, gint y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_screen_get_height_mm (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_icon_list (GdkWindow *window, GList *pixbufs)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_drag_status (GdkDragContext *context, GdkDragAction action, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_display_keyboard_ungrab (GdkDisplay *display, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_lower (GdkWindow *window)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean _gdk_windowing_window_queue_antiexpose (GdkWindow *window, GdkRegion *area)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_begin_move_drag (GdkWindow *window, gint button, gint root_x, gint root_y, guint32 timestamp)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_move_resize (GdkWindow *window, gint x, gint y, gint width, gint height)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkDisplay * gdk_display_open (const gchar *display_name)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_display_request_selection_notification (GdkDisplay *display, GdkAtom selection)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkColormap * gdk_screen_get_default_colormap (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_display_pointer_ungrab (GdkDisplay *display, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_set_background (GdkWindow *window, const GdkColor *color)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gint gdk_colormap_alloc_colors (GdkColormap *colormap, GdkColor *colors, gint ncolors, gboolean writeable, gboolean best_match, gboolean *success)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gint gdk_display_get_n_screens (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GType _gdk_pixmap_impl_get_type (void)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_icon (GdkWindow *window, GdkWindow *icon_window, GdkPixmap *pixmap, GdkBitmap *mask)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_resize (GdkWindow *window, gint width, gint height)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkFont * gdk_font_from_description_for_display (GdkDisplay *display, PangoFontDescription *font_desc)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_keymap_get_entries_for_keyval (GdkKeymap *keymap, guint keyval, GdkKeymapKey **keys, gint *n_keys)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+GdkWindow * gdk_window_lookup_for_display (GdkDisplay *display, GdkNativeWindow anid)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_drop_reply (GdkDragContext *context, gboolean ok, guint32 time)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkKeymap * gdk_keymap_get_for_display (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gchar * gdk_keyval_name (guint keyval)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_spawn_on_screen (GdkScreen *screen, const gchar *working_directory, gchar **argv, gchar **envp, GSpawnFlags flags, GSpawnChildSetupFunc child_setup, gpointer user_data, gint *child_pid, GError **error)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_keymap_get_entries_for_keycode (GdkKeymap *keymap, guint hardware_keycode, GdkKeymapKey **keys, guint **keyvals, gint *n_entries)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+gboolean gdk_display_supports_selection_notification (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+G_CONST_RETURN gchar * gdk_display_get_name (GdkDisplay *display)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_shape_combine_region (GdkWindow *window, GdkRegion *shape_region, gint offset_x, gint offset_y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+GdkColormap * gdk_screen_get_rgba_colormap (GdkScreen *screen)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
+
+void gdk_window_set_icon_name (GdkWindow *window, const gchar *name)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_window_get_root_origin (GdkWindow *window, gint *x, gint *y)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+void gdk_device_free_history (GdkTimeCoord **events, gint n_events)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+}
+
+gboolean gdk_device_get_history (GdkDevice *device, GdkWindow *window, guint32 start, guint32 stop, GdkTimeCoord ***events, gint *n_events)
+{
+	ENTER();
+	ASSERT();
+	LEAVE();
+	return 0;
+}
diff -Naur gtk+-2.10.14/gdk/xynth/gdkxynth.h gtk+-2.10.14-xynth/gdk/xynth/gdkxynth.h
--- gtk+-2.10.14/gdk/xynth/gdkxynth.h	1970-01-01 02:00:00.000000000 +0200
+++ gtk+-2.10.14-xynth/gdk/xynth/gdkxynth.h	2007-12-03 16:28:28.000000000 +0200
@@ -0,0 +1,45 @@
+
+#include <config.h>
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "gdk.h"
+#include "gdkwindow.h"
+
+#include "gdkregion-generic.h"
+
+#include "gdkinternals.h"
+#include "gdkalias.h"
+
+#define XYNTH_DEBUG 99
+
+#if XYNTH_DEBUG != 0
+#define DEBUG(fmt...) {\
+	printf(fmt);\
+	printf(" [%s (%s:%d)]\n", __FUNCTION__, __FILE__, __LINE__);\
+}
+#else
+#define DEBUG(fmt...) do { } while (0)
+#endif
+
+#if XYNTH_DEBUG == 99
+#define ENTER() DEBUG("Enter");
+#define LEAVE() DEBUG("Leave");
+#else
+#endif
+
+#define NIY()   {\
+	DEBUG("Not Implemented Yet");\
+}
+
+#define ASSERT() {\
+	DEBUG("This should not happen");\
+	assert(0);\
+}
+
+#define ERROR(fmt...) {\
+	DEBUG("ERROR" fmt); \
+	ASSERT(); \
+}
diff -Naur gtk+-2.10.14/gtk/Makefile.am gtk+-2.10.14-xynth/gtk/Makefile.am
--- gtk+-2.10.14/gtk/Makefile.am	2007-07-16 22:44:17.000000000 +0300
+++ gtk+-2.10.14-xynth/gtk/Makefile.am	2007-12-03 17:18:32.000000000 +0200
@@ -798,12 +798,14 @@
 libgtk_win32_2_0_la_SOURCES = $(gtk_c_sources)
 libgtk_quartz_2_0_la_SOURCES = $(gtk_c_sources)
 libgtk_directfb_2_0_la_SOURCES = $(gtk_c_sources)
+libgtk_xynth_2_0_la_SOURCES = $(gtk_c_sources)
 
 libgtk_x11_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_linux_fb_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_win32_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_quartz_2_0_la_LDFLAGS = $(libtool_opts)
 libgtk_directfb_2_0_la_LDFLAGS = $(libtool_opts)
+libgtk_xynth_2_0_la_LDFLAGS = $(libtool_opts)
 
 libgtk_x11_2_0_la_LIBADD = $(libadd)
 libgtk_linux_fb_2_0_la_LIBADD = $(libadd)
@@ -811,11 +813,12 @@
 libgtk_win32_2_0_la_DEPENDENCIES = $(gtk_def) $(gtk_win32_res)
 libgtk_quartz_2_0_la_LIBADD = $(libadd)
 libgtk_directfb_2_0_la_LIBADD = $(libadd)
+libgtk_xynth_2_0_la_LIBADD = $(libadd)
 
 if USE_WIN32
 libgtk_target_ldflags = $(gtk_win32_res_ldflag) $(gtk_win32_symbols)
 endif
-EXTRA_LTLIBRARIES = libgtk-x11-2.0.la libgtk-linux-fb-2.0.la libgtk-win32-2.0.la libgtk-quartz-2.0.la libgtk-directfb-2.0.la
+EXTRA_LTLIBRARIES = libgtk-x11-2.0.la libgtk-linux-fb-2.0.la libgtk-win32-2.0.la libgtk-quartz-2.0.la libgtk-directfb-2.0.la libgtk-xynth-2.0.la
 
 install-exec-hook: 
 if DISABLE_EXPLICIT_DEPS
