Index: test/awt/WidgetsDemo/WidgetsDemo.sh
===================================================================
RCS file: /cvs/kaffe/kaffe/test/awt/WidgetsDemo/WidgetsDemo.sh,v
retrieving revision 1.1
diff -u -r1.1 WidgetsDemo.sh
--- test/awt/WidgetsDemo/WidgetsDemo.sh	14 Jul 1998 14:20:05 -0000	1.1
+++ test/awt/WidgetsDemo/WidgetsDemo.sh	21 Aug 2006 13:20:10 -0000
@@ -1,2 +1,3 @@
 #! /bin/sh
-kaffe WidgetsDemo
+jikes +Pno-switchcheck +Pno-shadow +E -g -classpath .:../../../build-xynth/libraries/javalib/external/classpath/lib/glibj.zip -d . ./WidgetsDemo.java 
+kaffe -Xkaffe-xynth-awt WidgetsDemo
Index: README.xynth
===================================================================
RCS file: README.xynth
diff -N README.xynth
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ README.xynth	1 Jan 1970 00:00:00 -0000
@@ -0,0 +1,10 @@
+
+# developers/autogen.sh
+# mkdir build-xynth
+# cd build-xynth
+# ../configure --with-kaffe-xynth-awt --disable-gtk-peer --disable-gconf-peer
+# make
+# make install
+
+ Alper Akcan <distchx@yahoo.com>
+
